File: collector/src/main/java/com/bitmovin/analytics/stateMachines/PlayerStates.java
package com.bitmovin.analytics.stateMachines;

import com.bitmovin.analytics.data.ErrorCode;
import com.bitmovin.analytics.enums.AnalyticsErrorCodes;

public class PlayerStates
{
    public static PlayerState<Void> READY =
            new PlayerState<Void>() {
                @Override
                public void onEnterState(PlayerStateMachine machine, Void data) {}

                @Override
                public void onExitState(
                        PlayerStateMachine machine,
                        long elapsedTime,
                        PlayerState<?> destinationPlayerState) {}
            };
    public static PlayerState<Void> SOURCE_CHANGED =
            new PlayerState<Void>() {
                @Override
                public void onEnterState(PlayerStateMachine machine, Void data) {}

                @Override
                public void onExitState(
                        PlayerStateMachine machine,
                        long elapsedTime,
                        PlayerState<?> desintationPlayerState) {}
            };
    public static PlayerState<Void> STARTUP =
            new PlayerState<Void>() {
                @Override
                public void onEnterState(PlayerStateMachine machine, Void data) {
                    machine.videoStartTimeout.start();
                }

                @Override
                public void onExitState(
                        PlayerStateMachine machine,
                        long elapsedTime,
                        PlayerState<?> destinationPlayerState) {
                    machine.videoStartTimeout.cancel();
                    long elapsedTimeOnEnter = machine.getElapsedTimeOnEnter();
                    machine.addStartupTime(elapsedTime - elapsedTimeOnEnter);
                    if (destinationPlayerState == PlayerStates.PLAYING) {
                        long playerStartupTime = machine.getAndResetPlayerStartupTime();
                        for (StateMachineListener listener : machine.getListeners()) {
                            listener.onStartup(machine.getStartupTime(), playerStartupTime);
                        }
                        machine.setStartupFinished(true);
                    }
                }
            };
    public static PlayerState<Void> AD =
            new PlayerState<Void>() {
                @Override
                public void onEnterState(PlayerStateMachine machine, Void data) {}

                @Override
                public void onExitState(
                        PlayerStateMachine machine,
                        long elapsedTime,
                        PlayerState<?> destinationPlayerState) {}
            };
    public static PlayerState<Void> ADFINISHED =
            new PlayerState<Void>() {
                @Override
                public void onEnterState(PlayerStateMachine machine, Void data) {}

                @Override
                public void onExitState(
                        PlayerStateMachine machine,
                        long elapsedTime,
                        PlayerState<?> destinationPlayerState) {}
            };
    public static PlayerState<Void> BUFFERING =
            new PlayerState<Void>() {
                @Override
                public void onEnterState(PlayerStateMachine machine, Void data) {
                    machine.enableRebufferHeartbeat();
                    machine.rebufferingTimeout.start();
                }

                @Override
                public void onExitState(
                        PlayerStateMachine machine,
                        long elapsedTime,
                        PlayerState<?> desintationPlayerState) {
                    machine.disableRebufferHeartbeat();
                    for (StateMachineListener listener : machine.getListeners()) {
                        long elapsedTimeOnEnter = machine.getElapsedTimeOnEnter();
                        listener.onRebuffering(elapsedTime - elapsedTimeOnEnter);
                    }
                    machine.rebufferingTimeout.cancel();
                }
            };
    public static PlayerState<Void> ERROR =
            new PlayerState<Void>() {
                @Override
                public void onEnterState(PlayerStateMachine machine, Void data) {
                    machine.videoStartTimeout.cancel();
                    for (StateMachineListener listener : machine.getListeners()) {
                        listener.onError(machine.getErrorCode());
                    }
                }

                @Override
                public void onExitState(
                        PlayerStateMachine machine,
                        long elapsedTime,
                        PlayerState<?> desintationPlayerState) {
                    machine.setVideoStartFailedReason(null);
                }
            };
    public static PlayerState<Void> EXITBEFOREVIDEOSTART =
            new PlayerState<Void>() {
                @Override
                public void onEnterState(PlayerStateMachine machine, Void data) {
                    for (StateMachineListener listener : machine.getListeners()) {
                        listener.onVideoStartFailed();
                    }
                }

                @Override
                public void onExitState(
                        PlayerStateMachine machine,
                        long elapsedTime,
                        PlayerState<?> desintationPlayerState) {
                    machine.setVideoStartFailedReason(null);
                }
            };
    public static PlayerState<Void> PLAYING =
            new PlayerState<Void>() {
                @Override
                public void onEnterState(PlayerStateMachine machine, Void data) {
                    machine.enableHeartbeat();
                }

                @Override
                public void onExitState(
                        PlayerStateMachine machine,
                        long elapsedTime,
                        PlayerState<?> desintationPlayerState) {
                    for (StateMachineListener listener : machine.getListeners()) {
                        long elapsedTimeOnEnter = machine.getElapsedTimeOnEnter();
                        listener.onPlayExit(elapsedTime - elapsedTimeOnEnter);
                    }

                    machine.disableHeartbeat();
                }
            };
    public static PlayerState<Void> PAUSE =
            new PlayerState<Void>() {
                @Override
                public void onEnterState(PlayerStateMachine machine, Void data) {}

                @Override
                public void onExitState(
                        PlayerStateMachine machine,
                        long elapsedTime,
                        PlayerState<?> desintationPlayerState) {
                    for (StateMachineListener listener : machine.getListeners()) {
                        long elapsedTimeOnEnter = machine.getElapsedTimeOnEnter();
                        listener.onPauseExit(elapsedTime - elapsedTimeOnEnter);
                    }
                }
            };
    public static PlayerState<Void> QUALITYCHANGE =
            new PlayerState<Void>() {
                @Override
                public void onEnterState(PlayerStateMachine machine, Void data) {
                    machine.increaseQualityChangeCount();
                    if (!machine.isQualityChangeTimerRunning) {
                        machine.qualityChangeResetTimeout.start();
                    }
                }

                @Override
                public void onExitState(
                        PlayerStateMachine machine,
                        long elapsedTime,
                        PlayerState<?> destinationPlayerState) {
                    if (machine.isQualityChangeEventEnabled()) {
                        for (StateMachineListener listener : machine.getListeners()) {
                            listener.onQualityChange();
                        }
                    } else {
                        ErrorCode errorCode =
                                AnalyticsErrorCodes.ANALYTICS_QUALITY_CHANGE_THRESHOLD_EXCEEDED
                                        .getErrorCode();

                        for (StateMachineListener listener : machine.getListeners()) {
                            listener.onError(errorCode);
                        }
                    }
                }
            };
    public static PlayerState<Void> CUSTOMDATACHANGE =
            new PlayerState<Void>() {
                @Override
                public void onEnterState(PlayerStateMachine machine, Void data) {}

                @Override
                public void onExitState(
                        PlayerStateMachine machine,
                        long elapsedTime,
                        PlayerState<?> destinationPlayerState) {}
            };
    public static PlayerState<Void> AUDIOTRACKCHANGE =
            new PlayerState<Void>() {
                @Override
                public void onEnterState(PlayerStateMachine machine, Void data) {}

                @Override
                public void onExitState(
                        PlayerStateMachine machine,
                        long elapsedTime,
                        PlayerState<?> destinationPlayerState) {
                    for (StateMachineListener listener : machine.getListeners()) {
                        listener.onAudioTrackChange();
                    }
                }
            };
    public static PlayerState<Void> SUBTITLECHANGE =
            new PlayerState<Void>() {
                @Override
                public void onEnterState(PlayerStateMachine machine, Void data) {}

                @Override
                public void onExitState(
                        PlayerStateMachine machine,
                        long elapsedTime,
                        PlayerState<?> destinationPlayerState) {
                    for (StateMachineListener listener : machine.getListeners()) {
                        listener.onSubtitleChange();
                    }
                }
            };

    public static PlayerState<Void> SEEKING =
            new PlayerState<Void>() {
                @Override
                public void onEnterState(PlayerStateMachine machine, Void data) {
                    machine.setElapsedTimeSeekStart(machine.getElapsedTimeOnEnter());
                }

                @Override
                public void onExitState(
                        PlayerStateMachine machine,
                        long elapsedTime,
                        PlayerState<?> destinationPlayerState) {
                    for (StateMachineListener listener : machine.getListeners()) {
                        listener.onSeekComplete(elapsedTime - machine.getElapsedTimeSeekStart());
                    }
                    machine.setElapsedTimeSeekStart(0);
                }
            };
}

File: collector/src/main/java/com/bitmovin/analytics/stateMachines/PlayerStates.kt
package com.bitmovin.analytics.stateMachines

import android.util.Log
import com.bitmovin.analytics.dtos.ErrorCode
import com.bitmovin.analytics.dtos.SubtitleDto
import com.bitmovin.analytics.enums.AnalyticsErrorCodes

class PlayerStates {
    companion object {
        private const val TAG = "PlayerStates"

        @JvmField val READY = DefaultPlayerState<Void>("ready")

        @JvmField val SOURCE_CHANGED = DefaultPlayerState<Void>("source_changed")

        @JvmField val STARTUP =
            object : DefaultPlayerState<Void>("startup") {
                override fun onEnterState(
                    machine: PlayerStateMachine,
                    data: Void?,
                ) {
                    machine.videoStartTimeoutTimer.start()
                }

                override fun onExitState(
                    machine: PlayerStateMachine,
                    elapsedTime: Long,
                    durationInState: Long,
                    destinationPlayerState: PlayerState<*>,
                ) {
                    machine.videoStartTimeoutTimer.cancel()
                    machine.addStartupTime(durationInState)

                    if (destinationPlayerState === PlayerStates.PLAYING) {
                        // this exit marks a valid startup, which means we always
                        // need to have a valid videostartup_time
                        // there can be edge cases where we transition so fast that the durationInState is 0
                        // thus we add at least 1ms for these cases
                        if (durationInState == 0L) {
                            machine.addStartupTime(1)
                        }

                        val playerStartupTime = machine.getAndResetPlayerStartupTime()
                        machine.listeners.notify {
                            it.onStartup(machine, machine.startupTime, playerStartupTime)
                        }
                        machine.isStartupFinished = true
                    }
                }
            }

        @JvmField val AD =
            object : DefaultPlayerState<Void>("ad") {
                override fun onExitState(
                    machine: PlayerStateMachine,
                    elapsedTime: Long,
                    durationInState: Long,
                    destinationPlayerState: PlayerState<*>,
                ) {
                    machine.listeners.notify { it.onAd(machine, durationInState) }
                }
            }

        @JvmField val ADFINISHED = DefaultPlayerState<Void>("adfinished")

        @JvmField val BUFFERING =
            object : DefaultPlayerState<Void>("buffering") {
                override fun onEnterState(
                    machine: PlayerStateMachine,
                    data: Void?,
                ) {
                    machine.enableRebufferHeartbeat()
                    machine.bufferingTimeoutTimer.start()
                }

                override fun onExitState(
                    machine: PlayerStateMachine,
                    elapsedTime: Long,
                    durationInState: Long,
                    destinationPlayerState: PlayerState<*>,
                ) {
                    machine.disableRebufferHeartbeat()
                    machine.listeners.notify { it.onRebuffering(machine, durationInState) }
                    machine.bufferingTimeoutTimer.cancel()
                }
            }

        @JvmField val ERROR =
            object : DefaultPlayerState<ErrorCode>("error") {
                override fun onEnterState(
                    machine: PlayerStateMachine,
                    data: ErrorCode?,
                ) {
                    machine.videoStartTimeoutTimer.cancel()
                    if (data != null && !machine.shouldReportError(data.errorCode)) {
                        Log.i(TAG, "ErrorCode ${data.errorCode} already reported 5 times, not reporting this occurrence!")
                        return
                    }

                    machine.listeners.notify { it.onError(machine, data) }
                }

                override fun onExitState(
                    machine: PlayerStateMachine,
                    elapsedTime: Long,
                    durationInState: Long,
                    destinationPlayerState: PlayerState<*>,
                ) {
                    machine.videoStartFailedReason = null
                }
            }

        // This state is only used when EBVS or when we have a timeout during startup (as of 2025-08)
        @JvmField val VIDEOSTART_FAILED =
            object : DefaultPlayerState<Void>("videostart_failed") {
                override fun onEnterState(
                    machine: PlayerStateMachine,
                    data: Void?,
                ) {
                    machine.listeners.notify { it.onVideoStartFailed(machine, machine.startupTime) }
                }

                override fun onExitState(
                    machine: PlayerStateMachine,
                    elapsedTime: Long,
                    durationInState: Long,
                    destinationPlayerState: PlayerState<*>,
                ) {
                    machine.videoStartFailedReason = null
                }
            }

        @JvmField val PLAYING =
            object : DefaultPlayerState<Void>("playing") {
                override fun onEnterState(
                    machine: PlayerStateMachine,
                    data: Void?,
                ) {
                    // we reset the error limiter here since entering playing state means
                    // that the player is continuing playback and it is unlikely that we are in an error loop
                    // (which is usually during startup due to retries, or within the error state itself)
                    machine.resetIdenticalErrorReportingLimiter()
                    machine.enableHeartbeat()
                }

                override fun onExitState(
                    machine: PlayerStateMachine,
                    elapsedTime: Long,
                    durationInState: Long,
                    destinationPlayerState: PlayerState<*>,
                ) {
                    machine.listeners.notify { it.onPlayExit(machine, durationInState) }
                    machine.disableHeartbeat()
                }
            }

        @JvmField val PAUSE =
            object : DefaultPlayerState<Void>("pause") {
                override fun onExitState(
                    machine: PlayerStateMachine,
                    elapsedTime: Long,
                    durationInState: Long,
                    destinationPlayerState: PlayerState<*>,
                ) {
                    machine.listeners.notify { it.onPauseExit(machine, durationInState) }
                }
            }

        @JvmField val QUALITYCHANGE =
            object : DefaultPlayerState<Void>("qualitychange") {
                override fun onEnterState(
                    machine: PlayerStateMachine,
                    data: Void?,
                ) {
                    machine.qualityChangeEventLimiter.onQualityChange()
                }

                override fun onExitState(
                    machine: PlayerStateMachine,
                    elapsedTime: Long,
                    durationInState: Long,
                    destinationPlayerState: PlayerState<*>,
                ) {
                    if (machine.qualityChangeEventLimiter.isQualityChangeEventEnabled) {
                        machine.listeners.notify { it.onQualityChange(machine) }
                    } else {
                        val errorCode = AnalyticsErrorCodes.ANALYTICS_QUALITY_CHANGE_THRESHOLD_EXCEEDED.errorCode
                        machine.listeners.notify { it.onError(machine, errorCode) }
                    }
                }
            }

        @JvmField val CUSTOMDATACHANGE = DefaultPlayerState<Void>("customdatachange")

        @JvmField val AUDIOTRACKCHANGE =
            object : DefaultPlayerState<Void>("audiotrackchange") {
                override fun onExitState(
                    machine: PlayerStateMachine,
                    elapsedTime: Long,
                    durationInState: Long,
                    destinationPlayerState: PlayerState<*>,
                ) {
                    machine.listeners.notify { it.onAudioTrackChange(machine) }
                }
            }

        @JvmField val SUBTITLECHANGE =
            object : DefaultPlayerState<SubtitleDto>("subtitlechange") {
                override fun onExitState(
                    machine: PlayerStateMachine,
                    elapsedTime: Long,
                    durationInState: Long,
                    destinationPlayerState: PlayerState<*>,
                ) {
                    machine.listeners.notify { it.onSubtitleChange(machine, dataOnEnter) }
                }
            }

        @JvmField val SEEKING =
            object : DefaultPlayerState<Void>("seeking") {
                override fun onExitState(
                    machine: PlayerStateMachine,
                    elapsedTime: Long,
                    durationInState: Long,
                    destinationPlayerState: PlayerState<*>,
                ) {
                    machine.listeners.notify { it.onSeekComplete(machine, durationInState) }
                }
            }
    }
}
